Travis build hangs

The Travis build has reached `go test` and hung before any tests have run several times now before being terminated by Travis CI.

It's happened recently in:
[job 433357210](https://travis-ci.org/miekg/dns/jobs/433357210)—go 1.10—as part of #754, and
[job 433352922](https://travis-ci.org/miekg/dns/jobs/433352922)—go tip—as part of #751.

This failure is transient and the test will pass after being reset. This should be investigated.

This is an example of log output:
```
travis_time:start:32705ed3
$ go test $TESTS


No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received

The build has been terminated
```

I was on the right track with #774. I augmented the UDP code with logging in #761 and lo and behold:
```
2018/10/05 16:37:36 0xc0000a2618 in ShutdownContext
2018/10/05 16:37:36 0xc0000a2618 SetReadDeadline: 1970-01-01 00:00:01 +0000 UTC
2018/10/05 16:37:36 0xc0000a2618 SetReadDeadline: 2018-10-05 17:37:36.950566684 +0000 UTC m=+3663.040450987
2018/10/05 16:37:36 0xc0000a2618 ReadMsgUDP
```

```
panic: test timed out after 5m0s
goroutine 2498 [running]:
testing.(*M).startAlarm.func1()
	/home/travis/.gimme/versions/go1.11.1.linux.amd64/src/testing/testing.go:1296 +0x13a
created by time.goFunc
	/home/travis/.gimme/versions/go1.11.1.linux.amd64/src/time/sleep.go:172 +0x52
goroutine 1 [chan receive, 3 minutes]:
testing.(*T).Run(0xc00011e100, 0xadcf60, 0x17, 0xafa930, 0xc0000e9b01)
	/home/travis/.gimme/versions/go1.11.1.linux.amd64/src/testing/testing.go:879 +0x689
testing.runTests.func1(0xc00011e100)
	/home/travis/.gimme/versions/go1.11.1.linux.amd64/src/testing/testing.go:1119 +0xa9
testing.tRunner(0xc00011e100, 0xc0000e9d20)
	/home/travis/.gimme/versions/go1.11.1.linux.amd64/src/testing/testing.go:827 +0x163
testing.runTests(0xc0000ae380, 0xf04700, 0xba, 0xba, 0xc0000e9dc8)
	/home/travis/.gimme/versions/go1.11.1.linux.amd64/src/testing/testing.go:1117 +0x4ef
testing.(*M).Run(0xc000180000, 0x0)
	/home/travis/.gimme/versions/go1.11.1.linux.amd64/src/testing/testing.go:1034 +0x2ef
main.main()
	_testmain.go:598 +0x333
goroutine 2270 [semacquire, 3 minutes]:
sync.runtime_Semacquire(0xc00036ef48)
	/home/travis/.gimme/versions/go1.11.1.linux.amd64/src/runtime/sema.go:56 +0x39
sync.(*WaitGroup).Wait(0xc00036ef40)
	/home/travis/.gimme/versions/go1.11.1.linux.amd64/src/sync/waitgroup.go:130 +0xb5
github.com/miekg/dns.TestServerStartStopRace(0xc00011ec00)
	/home/travis/gopath/src/github.com/miekg/dns/server_test.go:894 +0x1cf
testing.tRunner(0xc00011ec00, 0xafa930)
	/home/travis/.gimme/versions/go1.11.1.linux.amd64/src/testing/testing.go:827 +0x163
created by testing.(*T).Run
	/home/travis/.gimme/versions/go1.11.1.linux.amd64/src/testing/testing.go:878 +0x651
goroutine 2396 [select, 3 minutes]:
github.com/miekg/dns.(*Server).ShutdownContext(0xc00011f700, 0xb756a0, 0xc00009e010, 0xc000000008, 0xc000560040)
	/home/travis/gopath/src/github.com/miekg/dns/server.go:486 +0x4de
github.com/miekg/dns.(*Server).Shutdown(0xc00011f700, 0xafbba0, 0xc00036ef40)
	/home/travis/gopath/src/github.com/miekg/dns/server.go:438 +0x7d
github.com/miekg/dns.TestServerStartStopRace.func1(0xc00036ef40, 0xc00011f700, 0xc00011ec00)
	/home/travis/gopath/src/github.com/miekg/dns/server_test.go:889 +0x67
created by github.com/miekg/dns.TestServerStartStopRace
	/home/travis/gopath/src/github.com/miekg/dns/server_test.go:887 +0xbd
goroutine 2395 [IO wait, 3 minutes]:
internal/poll.runtime_pollWait(0x7f1d010feba8, 0x72, 0x0)
	/home/travis/.gimme/versions/go1.11.1.linux.amd64/src/runtime/netpoll.go:173 +0x66
internal/poll.(*pollDesc).wait(0xc0008fab18, 0x72, 0xc000b5a700, 0x1, 0xc0004882a0)
	/home/travis/.gimme/versions/go1.11.1.linux.amd64/src/internal/poll/fd_poll_runtime.go:85 +0xe4
internal/poll.(*pollDesc).waitRead(0xc0008fab18, 0xc0006c2400, 0x200, 0x200)
	/home/travis/.gimme/versions/go1.11.1.linux.amd64/src/internal/poll/fd_poll_runtime.go:90 +0x4b
internal/poll.(*FD).ReadMsg(0xc0008fab00, 0xc0006c2400, 0x200, 0x200, 0xc0004882a0, 0x28, 0x28, 0x0, 0x0, 0x0, ...)
	/home/travis/.gimme/versions/go1.11.1.linux.amd64/src/internal/poll/fd_unix.go:243 +0x28e
net.(*netFD).readMsg(0xc0008fab00, 0xc0006c2400, 0x200, 0x200, 0xc0004882a0, 0x28, 0x28, 0xc00058b6e0, 0xf128a0, 0xf128a0, ...)
	/home/travis/.gimme/versions/go1.11.1.linux.amd64/src/net/fd_unix.go:214 +0xa1
net.(*UDPConn).readMsg(0xc0000a2610, 0xc0006c2400, 0x200, 0x200, 0xc0004882a0, 0x28, 0x28, 0x70, 0xbee602b438b73f13, 0xead8fb2bc, ...)
	/home/travis/.gimme/versions/go1.11.1.linux.amd64/src/net/udpsock_posix.go:59 +0xb5
net.(*UDPConn).ReadMsgUDP(0xc0000a2610, 0xc0006c2400, 0x200, 0x200, 0xc0004882a0, 0x28, 0x28, 0xc000b5ac30, 0x9d28d1, 0x432900, ...)
	/home/travis/.gimme/versions/go1.11.1.linux.amd64/src/net/udpsock.go:142 +0xb4
github.com/miekg/dns.(*loggingUDPConn).ReadMsgUDP(0xc0000a2618, 0xc0006c2400, 0x200, 0x200, 0xc0004882a0, 0x28, 0x28, 0x0, 0x0, 0x0, ...)
	/home/travis/gopath/src/github.com/miekg/dns/server.go:319 +0x1d3
github.com/miekg/dns.ReadFromSessionUDP(0xc0000a2618, 0xc0006c2400, 0x200, 0x200, 0x0, 0x0, 0xf07600, 0x465e66)
	/home/travis/gopath/src/github.com/miekg/dns/udp.go:42 +0x113
github.com/miekg/dns.(*Server).readUDP(0xc00011f700, 0xc0000a2618, 0x34630b8a000, 0x490e9d, 0xc00011f7cc, 0xc0ffffffff, 0x0, 0xc00011f7c4, 0xc000b5adc8)
	/home/travis/gopath/src/github.com/miekg/dns/server.go:760 +0x117
github.com/miekg/dns.(*defaultReader).ReadUDP(0xc0000a2630, 0xc0000a2618, 0x34630b8a000, 0xc00011f700, 0xc000b5ae10, 0x48d9be, 0xc00036f250, 0x48833c, 0x48ffc9)
	/home/travis/gopath/src/github.com/miekg/dns/server.go:170 +0x8e
github.com/miekg/dns.(*Server).serveUDP(0xc00011f700, 0xc0000a2618, 0x0, 0x0)
	/home/travis/gopath/src/github.com/miekg/dns/server.go:575 +0x295
github.com/miekg/dns.(*Server).ActivateAndServe(0xc00011f700, 0x0, 0x0)
	/home/travis/gopath/src/github.com/miekg/dns/server.go:424 +0x574
github.com/miekg/dns.RunLocalUDPServerWithFinChan.func1(0xc000950840, 0xc00011f700, 0xb77d80, 0xc0000a2610)
	/home/travis/gopath/src/github.com/miekg/dns/server_test.go:83 +0x39
created by github.com/miekg/dns.RunLocalUDPServerWithFinChan
	/home/travis/gopath/src/github.com/miekg/dns/server_test.go:82 +0x30d
```

Look at the address (0xc0000a2618) of the UDP conn blocked in `ReadFromSessionUDP` and the order of the `SetReadDeadline` calls.

--- a/server.go
+++ b/server.go
@@ -416,13 +416,12 @@ func (srv *Server) Shutdown() error {
 // to terminate.
 func (srv *Server) ShutdownContext(ctx context.Context) error {
 	srv.lock.Lock()
-	started := srv.started
-	srv.started = false
-	srv.lock.Unlock()
-
-	if !started {
+	if !srv.started {
+		srv.lock.Unlock()
 		return &Error{err: "server not started"}
 	}
+
+	srv.started = false
 
 	if srv.PacketConn != nil {
 		srv.PacketConn.SetReadDeadline(aLongTimeAgo) // Unblock reads
@@ -432,10 +431,10 @@ func (srv *Server) ShutdownContext(ctx context.Context) error {
 		srv.Listener.Close()
 	}
 
-	srv.lock.Lock()
 	for rw := range srv.conns {
 		rw.SetReadDeadline(aLongTimeAgo) // Unblock reads
 	}
+
 	srv.lock.Unlock()
 
 	if testShutdownNotify != nil {
@@ -666,13 +665,15 @@ func (srv *Server) serveDNS(w *response) {
 }
 
 func (srv *Server) readTCP(conn net.Conn, timeout time.Duration) ([]byte, error) {
-	if srv.isStarted() {
-		// If we race with ShutdownContext, the read deadline may
-		// have been set in the distant past to unblock the read
-		// below. We must not override it, otherwise we may block
-		// ShutdownContext.
+	// If we race with ShutdownContext, the read deadline may
+	// have been set in the distant past to unblock the read
+	// below. We must not override it, otherwise we may block
+	// ShutdownContext.
+	srv.lock.Lock()
+	if srv.started {
 		conn.SetReadDeadline(time.Now().Add(timeout))
 	}
+	srv.lock.Unlock()
 
 	l := make([]byte, 2)
 	n, err := conn.Read(l)
@@ -708,10 +709,12 @@ func (srv *Server) readTCP(conn net.Conn, timeout time.Duration) ([]byte, error)
 }
 
 func (srv *Server) readUDP(conn *net.UDPConn, timeout time.Duration) ([]byte, *SessionUDP, error) {
-	if srv.isStarted() {
+	srv.lock.Lock()
+	if srv.started {
 		// See the comment in readTCP above.
 		conn.SetReadDeadline(time.Now().Add(timeout))
 	}
+	srv.lock.Unlock()
 
 	m := srv.udpPool.Get().([]byte)
 	n, s, err := ReadFromSessionUDP(conn, m)

--- a/server.go
+++ b/server.go
@@ -669,11 +669,11 @@ func (srv *Server) readTCP(conn net.Conn, timeout time.Duration) ([]byte, error)
 	// have been set in the distant past to unblock the read
 	// below. We must not override it, otherwise we may block
 	// ShutdownContext.
-	srv.lock.Lock()
+	srv.lock.RLock()
 	if srv.started {
 		conn.SetReadDeadline(time.Now().Add(timeout))
 	}
-	srv.lock.Unlock()
+	srv.lock.RUnlock()
 
 	l := make([]byte, 2)
 	n, err := conn.Read(l)
@@ -709,12 +709,12 @@ func (srv *Server) readTCP(conn net.Conn, timeout time.Duration) ([]byte, error)
 }
 
 func (srv *Server) readUDP(conn *net.UDPConn, timeout time.Duration) ([]byte, *SessionUDP, error) {
-	srv.lock.Lock()
+	srv.lock.RLock()
 	if srv.started {
 		// See the comment in readTCP above.
 		conn.SetReadDeadline(time.Now().Add(timeout))
 	}
-	srv.lock.Unlock()
+	srv.lock.RUnlock()
 
 	m := srv.udpPool.Get().([]byte)
 	n, s, err := ReadFromSessionUDP(conn, m)

